### PR TITLE
Allow complex arrays in *Euclidean pairwise

### PR DESCRIPTION
**Pull request by dkarrasch**
Alternative to #219, without code duplication.

Closes #219.

### CODE REVIEW

**Comment by dkarrasch** [comment]
Anyone who would like to take a look? I've also included some `require_one_based_indexing` checks which used to be implicit but always necessary (otherwise matrix multiplications wouldn't work). Is that fine or should I do that separately to better track individual changes?

File: src/metrics.jl [hunk 1]
```diff
@@ -612,45 +612,49 @@ nrmsd(a, b) = NormRMSDeviation()(a, b)
 # SqEuclidean/Euclidean
 function _pairwise!(r::AbstractMatrix, dist::Union{SqEuclidean,Euclidean},
                     a::AbstractMatrix, b::AbstractMatrix)
+    require_one_based_indexing(r, a, b)
     m, na, nb = get_pairwise_dims(r, a, b)
-    mul!(r, a', b)
+    inplace = promote_type(eltype(r), typeof(oneunit(eltype(a))'oneunit(eltype(b)))) === eltype(r)
```
Comment (johnnychen94) [inline src/metrics.jl:617]: Leave a comment for this might be helpful for future readers.

**Review by johnnychen94 — APPROVED** [review]
*__looks__* good to me (I don't use complex numbers).

**Comment by juliohm** [comment]
I wonder if this generalization could also work with color types? For example pairwise between Array{Gray}. I missed this functionality yesterday trying to cluster a collection of pixels with a method that calls `pairwise(Euclidean(), colors, colors)`.

**Comment by johnnychen94** [comment]
No, because `abs2` is undefined for most colorants. More over, in ColorVectorSpace v0.10 release `abs2(::RGB)` is removed.

> I missed this functionality yesterday trying to cluster a collection of pixels with a method that calls pairwise(Euclidean(), colors, colors).

Isn't `ImageDistances` designed for this purpose?

**Comment by juliohm** [comment]
In the context of general machine learning some features are colors and
some others are not. So the clustering algorithm we have in GeoStats.jl
needs to call a generic pairwise distance not just image channels as
features. It would be nice to have it work in general. You mean that not
even Gray has a meaningful distance definition?

**Comment by johnnychen94** [comment]
> You mean that not even `Gray` has a meaningful distance definition?

See discussion in https://github.com/JuliaGraphics/ColorVectorSpace.jl/issues/157

And `pairwise(Euclidean(), colors, colors)` is already supported by ImageDistances:

```julia
julia> using ImageDistances

julia> x = rand(Gray{Float32}, 16, 2);

julia> pairwise(Euclidean(), x, x)
2×2 Matrix{Float32}:
 0.0      1.73782
 1.73782  0.0

julia> pairwise(Euclidean(), x, Float32.(x)) # mixed type
2×2 Matrix{Float32}:
 0.0      1.73782
 1.73782  0.0
```

This is off-topic so here's the last clarification in this PR; if you have further specific concerns please open issues in either ColorVectorSpace or ImageDistances.

**Comment by juliohm** [comment]
I find it suboptimal to depend on a separate package to handle this use case. ImageDistances.jl was created with a different purpose in mind. Namely, to define truly spatial distances that look at more than one pixel at a time like the Hausdorff distance. It was not created to be a general approach to compute distance between colors.

It would be interesting to revive this discussion in some other issue later. The original design in Images.jl shouldn´t be forgot where an image is simply an Array of color objects. If we keep developing image-specific packages like ImageDistances.jl with a different picture in mind, we will create an isolated ecosystem that only works in specific use cases. The discussion around abs2 is controversial with many different opinions and we should reconsider a generic treatment of colors if that is possible.

Again I started ImageDistances.jl with a very specific set of distances in mind, the image-specific distances that cannot be expressed in terms of pixel-wise operations.

**Comment by dkarrasch** [comment]
Since this is already adding functionality, shall we have a separate discussion about the image-related issue? The main question seems to be what to do about the `abs2` colwise reduction, right? One possible solution could be to replace that by some `_abs2`, which defaults to `abs2` but could be overloaded by other packages for custom types.

Since there seem to be no objections, I'll go and merge this PR, but wait for other recent PRs before a release.

**Pull request merged**